### DISCUSSION TITLE
instantiating dynamic class inheriting from abstract class results on a "object has no attribute '_sa_instance_state'" error

### DISCUSSION THREAD

**Discussion by flixman** [sqlalchemy/sqlalchemy#7675] [Usage Questions] [answered]
Hi all! This is the first time I "discuss" on github. Please excuse me if this is not the proper place for my request. I am working on a project to read data from a family of sensors, and different sensors produce different data fields. As I want the project to be general, I have specified the fields for some of the sensors on a file and I want to create the tables for them. To that end, I am doing the following:

```python
class _Reading(db.Model):
    __abstract__ = True
    
    tstamp = db.Column('tstamp', db.DateTime, primary_key=True, autoincrement=False)

def solar_object_init(self, **entries):
    # entries can contain more kwargs than expected at table creation time. By updating __dict__ we get them ignored.
    self.__dict__.update(entries)

def _mapping_for_reading(classname: str, record: list[dict[str, str]]) -> Type:
    clsdict = {"__tablename__": classname.lower()}
    for entry in record:
        try:
            clsdict[entry['tag']] = db.Column(eval(entry["dbmodel"]))  # pylint: disable=W0123
        except KeyError:
            pass

    _cls = type(classname, (_Reading,), clsdict)
    _cls.__init__ = solar_object_init
    return _cls

Solar = {}
for record in solar_records_transferred:
    Solar[record] = _mapping_for_reading(f'Solar_{record}', solar_records[record])
```
When the previous code gets executed, when the module is loaded, I get in the postgresql backend the tables for "solar_xxxxx" classes (so that seems to be going ok). However, the moment I try to insert a new record, which I do like

```python
# reading['packet_code'] and reading['values'] have proper values for the table name and the values initialized before.
reading_obj = Solar[reading['packet_code']](**reading['values'])
```

I get the following error:
`sqlalchemy.orm.exc.UnmappedInstanceError: Class 'app.models.Solar_T040504' is mapped, but this instance lacks instrumentation.  This occurs when the instance is created before sqlalchemy.orm.mapper(app.models.Solar_T040504) was called.`

I am completely clueless about what might be wrong in my code. Can somebody give me a hint?

**Comment by zzzeek** [comment, marked as answer]
Hey there -

OK that's a pretty obscure error message to get, and in this case it's misleading, because the way it determines that condition is that it sees that the `__init__` method didn't do what we expected it to do.    SQLAlchemy applies a transparent decorator to your class' `__init__` method that does some important startup stuff.

short answer is don't monkeypatch `__init__`, send it in your clsdict:

```py
  clsdict["__init__"] = solar_object_init
  _cls = type(classname, (_Reading,), clsdict)
```

**Reply by flixman** [reply]
Thank you!!! :-)

**Reply by CaselIT** [reply]
since sqlalchemy has a metaclass in place we could raise on set of `__init__` ?

I guess it's not a common case though

**Reply by zzzeek** [reply]
sqlalchemy didnt originally use a metaclass and all the imperative mapping cases as well as `@mapped` and all that also dont use metaclasses.

this is a pretty weird edge case so i dont think we need to worry too hard about it

**Reply by CaselIT** [reply]
I agree.
